src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,186 @@ mod type_keyword {}
+/// # Examples
+///
+/// ## Marking elements as `unsafe`
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
+/// }
+///
+/// struct S;
+///
+/// impl S {
+///     unsafe fn unsafe_method_on_struct() {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
 ///
+/// ```rust
+/// unsafe trait UnsafeTrait {}
+/// ```
+///
+/// Since `unsafe fn` and `unsafe trait` indicate that there is a safety
+/// contract that the compiler cannot enfore, documenting it is important. The
+/// standard library has many examples of this, like the following which is an
+/// extract from [`Vec::set_len`]. The `# Safety` section explains the contract
+/// that must be fulfilled to safely call the function.
+///
+/// ```rust,ignore (stub-to-show-doc-example)
+/// /// Forces the length of the vector to `new_len`.
+/// ///
+/// /// This is a low-level operation that maintains none of the normal
+/// /// invariants of the type. Normally changing the length of a vector
+/// /// is done using one of the safe operations instead, such as
+/// /// `truncate`, `resize`, `extend`, or `clear`.
+/// ///
+/// /// # Safety
+/// ///
+/// /// - `new_len` must be less than or equal to `capacity()`.
+/// /// - The elements at `old_len..new_len` must be initialized.
+/// pub unsafe fn set_len(&mut self, new_len: usize)
+/// ```
+///
+/// ## Using `unsafe {}` blocks and `impl`s
+///
+/// Performing `unsafe` operations requires an `unsafe {}` block:
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// /// Dereference the given pointer.
+/// ///
+/// /// # Safety
+/// ///
+/// /// `ptr` must be aligned and must not be dangling.
+/// unsafe fn deref_unchecked(ptr: *const i32) -> i32 {
+///     *ptr
+/// }
+///
+/// let a = 3;
+/// let b = &a as *const _;
+/// // SAFETY: `a` has not been dropped and references are always aligned,
+/// // so `b` is a valid address.
+/// unsafe { assert_eq!(*b, deref_unchecked(b)); };
+/// ```
+///
+/// Traits marked as `unsafe` must be [`impl`]emented using `unsafe impl`. This
+/// makes a guarantee to other `unsafe` code that the implementation satisfies
+/// the trait's safety contract. The [Send] and [Sync] traits are examples of
+/// this behaviour in the standard library.
+///
+/// ```rust
+/// /// Implementors of this trait must guarantee an element is always
+/// /// accessible with index 3.
+/// unsafe trait ThreeIndexable<T> {
+///     fn three(&self) -> &T;
+///     /// This method will not perform any bound check before accessing the
+///     /// fourth element and is not guaranteed to panic if it is not.
+///     unsafe fn three_unchecked(&self) -> &T;
```

Comment:
Why is this method `unsafe`? If "an element is always accessible with index 3" then there is no possible way to cause unsoundness here, so the `unsafe` is overkill here.

  - (which makes me think it would be good to mention somewhere that "overkill" `unsafe`, despite valid, defeats the purpose of `unsafe`; it should thus be avoided (mentioning that functions such as `remove_dir_all()` are not marked `unsafe` since they are not memory unsafe, despite them being dangerous)).

___

I suggest making the `unchecked` variant non-`unsafe`, or changing the contract of the trait to allow fallible indexing

Reply:
That's because unsafe version does not perform bound check, like `get_unchecked`. It should be unsafe like how `get_unchecked` is unsafe.

Reply:
The trait currently has a safety invariant attached: that of the indexing operation being infallible, which `get_unchecked` has not (in other words, given the safety invariant of the trait, implementing the trait for `[T]` slices would violate the contract); that's what allows it to offer a non-`unsafe` `get_unchecked`.
It could be implemented for `[T; N] where N ≥ 3`, or for `(_, _, T, ...)`, where bound-checking the indexing operation can be skipped, without having to be `unsafe`.

  - [Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5421d0390186b9fe8439184ea2a28d7e)

I like that example because it shows the benefits / the power of `unsafe trait`s: by restricting the subset of applicable types (and thus inputs), you can make a generally `unsafe` operation no longer be so.

Reply:
I can remove the `unsafe fn three_unchecked` if it clutters the example while not providing value.

It is surprisingly hard to find a good example for an `unsafe trait` with an `unsafe fn` inside that is not contrived and/or very long and/or exceedingly technical (like showing `GlobalAlloc` would be).

Reply:
Removed the `unsafe fn tree_unchecked`, I hope the example makes more sense now, in conjunction with the remark from @rylev.